File: LeetCode/Tag/Binary Search Tree/js/897.increasing-order-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var increasingBST = function(root) {
    let resultNode,currentNode;

    function traverseTree(node){
        if(node==null)
            return;
        traverseTree(node.left);
        addToFinalTree(node.val);
        traverseTree(node.right);
    }

    function addToFinalTree(val){
        const newNode = new TreeNode(val,null,null);
        if(resultNode==null){
            currentNode = newNode;
            resultNode = currentNode;
        }
        else{
            currentNode.right = newNode;
            currentNode = currentNode.right;
        }
    }

    traverseTree(root);

    return resultNode;
};

File: LeetCode/Tag/Depth-first-Search/cpp/897.increasing-order-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<TreeNode*, TreeNode*> solve(TreeNode* root) {
        TreeNode *head = root, *tail = root;

        if (root->left) {
            auto res = solve(root->left);
            root->left = NULL;
            res.second->right = root;
            
            head = res.first;
        }

        if (root->right) {
            auto res = solve(root->right);
            root->right = res.first;
            
            tail = res.second;
        }
        return {head, tail};
    }

    TreeNode* increasingBST(TreeNode* root) {
        return solve(root).first;
    }
};